### ISSUE TITLE
Runner emits warnings on batched annex commands

### ISSUE THREAD

**Issue by bpoldrack** [datalad/datalad#6431]
I am using `git annex registerurl` in batch mode. Apparently there's one more line of output than expected (namely some `registerurl (ok)`  at the end). This results in the runner emitting a WARNING about remaining, unprocessed output.
Either that means, that batch commands need to change how they call the runner to make it clear in some way that this is expected. But I think of that as a maintenance burden as it requires to specifically find out about every annex command how exactly it behaves in that regard in batch mode (and in what annex version).
Or it should not be a WARNING to begin with. I lean towards that, since a user can't do anything about it anyway. What do you think, @christian-monch ?

**Comment by christian-monch** [comment]
The warning is emitted in the situation where `git-annex wrote` a non-terminated line to stdout and exited. The current implementation of `BatchedCommand` relies on the assumption that each input line is acknowledged with a result that consists of one or more complete output lines. Therefore the warning when an unterminated line was read when the command exits.

I am not sure about the significance of the output, i.e. `registerurl (ok)`. If it is an expected result, we must assume that git-annex sometimes terminates its output lines by exiting. If this is the case, we might add a flag to BatchedCommand to handle this state. Otherwise we can probably ignore the output and get rid of the warning, or demote it to debug-level.

Can you give an example of a successful interaction with `git-annex-registerurl` in batch mode?

**Comment by yarikoptic** [comment]
Oh, according to https://git-annex.branchable.com/git-annex-registerurl/ there is no --json mode which should ideally be the one for any programmable interface (attn @joeyh).

**Comment by christian-monch** [comment]
Using git-annex version `8.20211231`, I am surprised that it works at all. My version does not acknowledge valid input, so `BatchedCommand.__call__()` should actually wait forever.

**Comment by bpoldrack** [comment]
> Using git-annex version 8.20211231, I am surprised that it works at all.

I didn't even implement it myself. I am using `Batched RegisterUrl` from `addurls`.

**Comment by christian-monch** [comment]
@bpoldrack checked the code and `BatchedCommand` does the right thing, if it is instantiated with an `output_proc` that returns a result without reading anything from `stdout` of the command. This is the case in `BatchedRegisterUrl`. Therefore data is accumulating on stdout. `BatchedCommand` warns about that because it does not expect unprocessed output, which often indicates an error.

Here are some possible solutions (either in isolation or combined):

1. Demoting the warning to debug
2. Removing the warning altogether
3. Adding a flag to indicate that unprocessed output is expected
4. Modify the `output_proc` used by `BatchedRegisterUrl`, i.e. `BatchedRegisterUrl` to consume the expected output.

I guess option 4. would be the cleanest approach. It would probably trigger a failure should `git-annex` change its protocol in the future, which is positive because it decreases the probability of silently failing

**Comment by yarikoptic** [comment]
FWIW filed TODO https://git-annex.branchable.com/todo/--json_for_registerurl/

**Comment by bpoldrack** [comment]
@christian-monch 

Option 4 sounds cleanest indeed. I currently lean towards 1) in addition, though. That's because that message is relevant to devs only. Warning the user about it does no good, particularly if ultimately nothing went wrong really. Looks like problem, but isn't and even if it was the user couldn't do anything about it.

**Comment by christian-monch** [comment]
I think that if we use option 4, we should not demote the warning, because in this case, it would indicate that something has changed in the behavior of `git annex registerurl`. 

To be clear: I would only leave the warning in the code if we do modify the code to consume the output. Because then we would not expect any unhandled data unless the behavior of an external command changes. And such a change would be something we might want to see as early as possible. 

Generally, I think that warnings are only partially directed to the user. They indicate that something unexpected happened, and if an error occurs, that unexpected event might be the cause. A warning should at some point end up at a developer's desk. The user is indeed not concerned with the content of the warning, and can/should treat it as opaque information that can be included in a potential support request. It might simplify support because the information is already present and does not have to be gathered by additional measures, e.g. ask a user to rerun with debug output enabled. 

BTW: do we have a guiding document that describes the purposes and use of errors-, warnings-, info-, etc. levels? 

Action: I will either implement option 4 and leave it at warning-level, or demote it to debug-level

**Comment by bpoldrack** [comment]
> ... Generally, I think that warnings are only partially directed to the user. They indicate that something unexpected happened, and if an error occurs, that unexpected event might be the cause. A warning should at some point end up at a developer's desk.

True. The issue with that is that we still don't properly separate logging from user facing output. We should and if we did, I'd completely agree. But with the status quo I am a bit torn.

**Comment by joeyh** [comment]
--json is now implemented. This also involved changing the output in --batch mode.
